### PR TITLE
[PORT/Revival] Custom say emotes and Emphasis

### PR DESCRIPTION
**Pull request by Bobbanz1**
## About The Pull Request
Ports this PR from NSV13
[#2170](https://github.com/BeeStation/NSV13/pull/2170)
However that PR is a port from here so it's technically a revival of #5897 rather than a port?

Ports from Whitesands
[#496](https://github.com/Whitesands13/Whitesands/pull/496)

## Why It's Good For The Game
Allows for better Roleplaying over the Radio and in person with other characters.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

Will have to add this later but for now have this image from NSV13 of it working

200263401-e19f782b-3f3b-49b4-8d9e-9a73d63d588b

</details>

## Changelog
:cl:Kapu1178, Kevinz, XSlayer300, Ryll/Shaps, Seris02, GoldenAlpharex, Shadowtail117
add: Speech Format Help verb under the OOC tab.
add: Text emphasis. Surrounding a block of text with underscores will italicize it and surrounding it with two asterisks will bold it
add: Emotes that start with an apostrophe wont format weirdly
add: You can now have custom say mods, and can use them to emote over radio. say";screeches-HELP SECURITY!" results in [Common] Joe screeches, "HELP SECURITY!"
/:cl:

### CODE REVIEW

**Comment by EvilDragonfiend** [comment]
personally, I don't like `;screams*HELP!` format because you would expect a speech after `;` and expect an emotion after `*`. these should be reversed.

**Comment by Bobbanz1** [comment]
> personally, I don't like `;screams*HELP!` format because you would expect a speech after `;` and expect an emotion after `*`. these should be reversed.
> Not only me but also other players have the same issue. some people did `;help!*screams` then a radio got `Clownie HELP!, "screams."`

There is a third method of doing this

`;screams, "HELP ME!"*`

I can see about trying to figure out how to make it work through some different format however I’d need an agreed upon format that would be acceptable to all.

**Comment by PowerfulBacon** [comment]
I like bold and italics, not a fan of the underlined text since its not commonly used in conversational settings in order to provide emphasis (bold is usually used instead).

File: code/game/say.dm [hunk 1]
```diff
@@ -111,24 +118,38 @@ GLOBAL_LIST_INIT(freqtospan, list(
...
+/// Transforms the speech emphasis mods from [/atom/movable/proc/say_emphasis] into the appropriate HTML tags. Includes escaping.
+#define ENCODE_HTML_EMPHASIS(input, char, html, varname) \
+	var/static/regex/##varname = regex("(?<!\\\\)[char](.+?)(?<!\\\\)[char]", "g");\
+	input = varname.Replace_char(input, "<[html]>$1</[html]>")
+
+/// Scans the input sentence for speech emphasis modifiers, notably |italics|, +bold+, and _underline_ -mothblocks
+/atom/proc/say_emphasis(input)
+	ENCODE_HTML_EMPHASIS(input, "\\|", "i", italics)
+	ENCODE_HTML_EMPHASIS(input, "\\+", "b", bold)
+	ENCODE_HTML_EMPHASIS(input, "_", "u", underline)
```
Comment (PowerfulBacon) [inline code/game/say.dm]: I hate the + and |, bold should be * I think or the same as markdown \*\***bold**\*\*.

If the problem is that math may be picked up like "8 * 4 + 5 * 2", then we can just do what discord does and only allow it if there is no spaces between it.
```regex
\*\*[^ ](.*)[^ ]\*\*
```

I don't mind having both for people who play on multiple servers, but I think \* should either be bold, or use markdown for italics and boldness.

Comment (PowerfulBacon) [inline code/game/say.dm]: Also I don't think we should allow people to talk in underlines, since bold provides the emphasis that underlined would need and it kind of ends up looking more like a hyperlink than some important text.
image
If other maintainers disagree with me (I've asked in dev channel), then I'll resolve this.

**Review by PowerfulBacon — CHANGES_REQUESTED** [review]

**Review by Archanial — DISMISSED** [review]
Remove underline entirely
I'll review code in a bit

**Comment by Bobbanz1** [comment]
> Remove underline entirely
> I'll review code in a bit

I’ll do that upon getting home

File: code/game/say.dm (hunk 1 above)
Comment (Bobbanz1) [reply code/game/say.dm]: Underline has been removed

**Comment by Bobbanz1** [comment]
> Remove underline entirely I'll review code in a bit

Done

**Comment by Bobbanz1** [comment]
GOD DAMN IT NO

**Comment by Bobbanz1** [comment]
RIGHT BACK TO NORMAL

File: code/datums/emotes.dm [hunk 2]
```diff
@@ -201,3 +203,16 @@
 				ghost.show_message("[FOLLOW_LINK(ghost, src)] [ghost_text]")
 
 	visible_message(text, visible_message_flags = list(CHATMESSAGE_EMOTE = TRUE))
+
+/**
+ * Returns a boolean based on whether or not the string contains a comma or an apostrophe,
+ * to be used for emotes to decide whether or not to have a space between the name of the user
```
Comment (itsmeow) [inline code/datums/emotes.dm:209]: I'm really confused in what case you would NOT want a space

File: code/game/say.dm [hunk 3]
```diff
@@ -111,24 +118,29 @@ GLOBAL_LIST_INIT(freqtospan, list(
...
+	var/say_mod = message_mods[MODE_CUSTOM_SAY_EMOTE]
+	if(!say_mod)
+		say_mod = say_mod(input, message_mods)
+
 	var/spanned = attach_spans(input, spans)
-	return "[say_mod(input, message_mods)], \"[spanned]\""
+	return "[say_mod], \"[spanned]\""
+
+/atom/proc/say_emphasis(message, var/list/ignore = list())
+	var/regex/markup
```
Comment (itsmeow) [inline code/game/say.dm:130]: can you document what the hell this proc does, because I do not understand by looking at it

File: code/modules/mob/say.dm [hunk 4]
```diff
@@ -101,6 +101,19 @@
 ///The amount of items we are looking for in the message
 #define MESSAGE_MODS_LENGTH 6
 
+/mob/proc/check_for_custom_say_emote(message, list/mods)
```
Comment (itsmeow) [inline code/modules/mob/say.dm:116]: add a comment for what this does as well, it's a bit confusing

**Review by itsmeow — CHANGES_REQUESTED** [review]

File: code/game/say.dm (hunk 3 above)
Comment (Bobbanz1) [reply code/game/say.dm:130]: Done to the best of my current knowledge of it

File: code/modules/mob/say.dm (hunk 4 above)
Comment (Bobbanz1) [reply code/modules/mob/say.dm:116]: Done to the best of my current knowledge of it, might not be a useful comment but I tried my best.

File: code/datums/emotes.dm (hunk 2 above)
Comment (Bobbanz1) [reply code/datums/emotes.dm:209]: Removed the Proc and made it the way it was previously

Comment (Kapu1178) [reply code/datums/emotes.dm:209]: *me's eyeballs FUCKING EXPLODE gets spit out as `Kapu1178 's eyeballs FUCKING EXPLODE` without this.

Comment (Bobbanz1) [reply code/datums/emotes.dm:209]: Ooookay time to reinstall it then!

Comment (itsmeow) [reply code/datums/emotes.dm:209]: I just want someone to explain

Comment (itsmeow) [reply code/datums/emotes.dm:209]: Oh kapu explained. i thought it was a meme

**Review by itsmeow — DISMISSED** [review]
Good improvement!

**Comment by Pirill** [comment]
I'm... quite curious why the maints would like to remove underline as an option rather than leave it there just in case anyone wants to use it... This is a personal thing, of course, but I primarily use underline for emphasis on servers that have this feature so it just feels like an odd arbitrary removal of options, that doesn't really hurt anyone by being there, I dunno.

**Review by PowerfulBacon — DISMISSED** [review]
Briefly skimmed through the code, I don't see anything majorly wrong.
Very good changes to the control characters. I am satisfied.

**Review by Crossedfall — DISMISSED** [review]

File: code/__DEFINES/say.dm [hunk 5]
```diff
@@ -41,7 +41,9 @@
 #define MODE_VOCALCORDS "cords"
 #define MODE_KEY_VOCALCORDS "x"
 
-#define MODE_SING "%"
+#define MODE_SING "sing"
```
Comment (Archanial) [inline code/__DEFINES/say.dm:44]: why is this changed? Does that mean you can't say "sing" without singing?

File: code/datums/emotes.dm [hunk 6]
```diff
@@ -87,11 +87,13 @@
 
 	user.log_message(msg, LOG_EMOTE)
 
+	var/space = should_have_space_before_emote(html_decode(msg)[1]) ? " " : ""
```
Comment (Archanial) [inline code/datums/emotes.dm]:
```suggestion
	var/space = should_have_space_before_emote(html_decode(msg)[1]) ? " " : null
```

File: code/datums/emotes.dm [hunk 7]
```diff
@@ -201,3 +203,16 @@
...
+ * to be used for emotes to decide whether or not to have a space between the name of the user
+ * and the emote.
+ *
+ * Requires the message to be HTML decoded beforehand. Not doing it here for performance reasons.
+ *
+ * Returns TRUE if there should be a space, FALSE if there shouldn't.
+ */
+/proc/should_have_space_before_emote(string)
+	var/static/regex/no_spacing_emote_characters = regex(@"(,|')")
+	return no_spacing_emote_characters.Find(string) ? FALSE : TRUE
```
Comment (Archanial) [inline code/datums/emotes.dm]:
```suggestion
	return !no_spacing_emote_characters.Find(string)
```

File: code/modules/mob/mob.dm [hunk 8]
```diff
@@ -181,7 +181,7 @@
 	to_chat(src, msg, avoid_highlighting = avoid_highlighting)
 
 
-/atom/proc/visible_message(message, self_message, blind_message, vision_distance = DEFAULT_MESSAGE_RANGE, list/ignored_mobs, list/visible_message_flags, allow_inside_usr = FALSE)
+/atom/proc/visible_message(message, self_message, blind_message, vision_distance = DEFAULT_MESSAGE_RANGE, list/ignored_mobs, list/visible_message_flags, allow_inside_usr = FALSE, separation = " ")
```
Comment (Archanial) [inline code/modules/mob/mob.dm:184]: I am unsure if the default should be space. Doesn't this impact other cases than emotes etc?
Also this proc should handle null separator

File: code/modules/mob/say.dm [hunk 9]
```diff
@@ -101,6 +101,34 @@
...
+		return message
+	if(findtext(message, " ", customsaypos + 1, customsaypos + 2))
+		messagetextpos = 2
+	if(is_banned_from(ckey, "Emote"))
+		return copytext(message, customsaypos + messagetextpos)
+	mods[MODE_CUSTOM_SAY_EMOTE] = lowertext(copytext_char(message, 1, customsaypos))
+	message = copytext(message, customsaypos + messagetextpos)
+	if(!message)
+		mods[MODE_CUSTOM_SAY_ERASE_INPUT] = TRUE
+		message = "an interesting thing to say"
```
Comment (Archanial) [inline code/modules/mob/say.dm]: Are you sure it's a good idea to force people to say this?
Can you show any screenshots of how this looks?

**Review by Archanial — DISMISSED** [review]

File: code/modules/mob/say.dm (hunk 9 above)
Comment (Kapu1178) [reply code/modules/mob/say.dm]: its a debug tool. if you see this the yame is broken.

File: code/__DEFINES/say.dm (hunk 5 above)
Comment (Kapu1178) [reply code/__DEFINES/say.dm:44]: we went over this when i made the original pr. mode != key. its a code only thing to make its purpose clear.

File: code/datums/emotes.dm (hunk 6 above)
Comment (Kapu1178) [reply code/datums/emotes.dm]: why? typefucking for no reason makes bad code.

File: code/modules/mob/say.dm (hunk 9 above)
Comment (EvilDragonfiend) [reply code/modules/mob/say.dm]: Yeah I noticed this is to track when someone says this, but I think you need to put a stacktrace when they say this.

File: code/__DEFINES/say.dm (hunk 5 above)
Comment (Archanial) [reply code/__DEFINES/say.dm:44]: it might come as a surprise but I do not remember every answer to my every comment ever

File: code/datums/emotes.dm (hunk 6 above)
Comment (Archanial) [reply code/datums/emotes.dm]: because no separator (space) should be no separator not an empty string.
Everything in dm is nullable so I don't get the "typefucking" comment.

File: code/modules/mob/say.dm (hunk 9 above)
Comment (Archanial) [reply code/modules/mob/say.dm]: Stacktrace is better and will help us fix any bugs later

File: code/modules/mob/mob.dm (hunk 8 above)
Comment (PowerfulBacon) [reply code/modules/mob/mob.dm:184]: from the looks of the code, the seperation is just the character that comes between the name and the message. The default was space before, so there should be no behavioural changes from what we have now that isn't intended to be changed by this PR.
The only situation where you would want this as something like null, would be '*me 's suit sensors flicker out.'

File: code/modules/mob/say.dm (hunk 9 above)
Comment (PowerfulBacon) [reply code/modules/mob/say.dm]: I agree that this should either CRASH if we are expecting a message to be passed in (and then !message has to be checked every time this proc is called) or we expect message to be nullable and just return empty string or nullable and accept the input.

**Comment by PowerfulBacon** [comment]
Resolve arch's reviews for merge

**Review by itsmeow — APPROVED** [review]
